backends/aerospike.go

```diff
@@ -13,59 +15,111 @@ import (
 const setName = "uuid"
 const binValue = "value"
 
-type Aerospike struct {
+type AerospikeDB interface {
+	NewUuidKey(namespace string, key string) (*as.Key, error)
+	Get(key *as.Key) (*as.Record, error)
+	Put(key *as.Key, value string, ttlSeconds int) error
+}
+
+type AerospikeDBClient struct {
+	client *as.Client
+}
+
+func (db AerospikeDBClient) Get(key *as.Key) (*as.Record, error) {
+	rec, err := db.client.Get(nil, key, binValue)
+	if err != nil {
+		return nil, formatAerospikeError(err, "GET")
+	}
+	return rec, nil
+}
+
+func (db AerospikeDBClient) Put(key *as.Key, value string, ttlSeconds int) error {
+	bins := as.BinMap{binValue: value}
+	policy := &as.WritePolicy{Expiration: uint32(ttlSeconds)}
+
+	err := db.client.Put(policy, key, bins)
+
+	return formatAerospikeError(err, "PUT")
+}
+
+func (db *AerospikeDBClient) NewUuidKey(namespace string, key string) (*as.Key, error) {
+	asKey, err := as.NewKey(namespace, setName, key)
+	if err != nil {
+		return nil, err
+	}
+	return asKey, nil
+}
+
+type AerospikeBackend struct {
 	cfg     config.Aerospike
-	client  *as.Client
+	client  AerospikeDB
 	metrics *metrics.Metrics
 }
 
-func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *Aerospike {
+func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *AerospikeBackend {
+	if cfg.Host == "" {
+		log.Fatalf("Cannot connect to empty Aerospike host")
+	}
+	if cfg.Port <= 0 {
+		log.Fatalf("Cannot connect to Aerospike host at port %d", cfg.Port)
+	}
 	client, err := as.NewClient(cfg.Host, cfg.Port)
 	if err != nil {
-		log.Fatalf("Error creating Aerospike backend: %v", err)
-		panic("Aerospike failure. This shouldn't happen.")
+		log.Fatalf("Error creating Aerospike backend: %v", formatAerospikeError(err, "NewAerospikeBackend"))
+		panic("AerospikeBackend failure. This shouldn't happen.")
 	}
 	log.Infof("Connected to Aerospike at %s:%d", cfg.Host, cfg.Port)
 
-	return &Aerospike{
+	return &AerospikeBackend{
 		cfg:     cfg,
-		client:  client,
+		client:  &AerospikeDBClient{client},
 		metrics: metrics,
 	}
 }
 
-func (a *Aerospike) Get(ctx context.Context, key string) (string, error) {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Get(ctx context.Context, key string) (string, error) {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, key)
 	if err != nil {
-		return "", err
+		return "", formatAerospikeError(err, "GET")
```

Comment:
I know we discussed offline that having the caller for `GET` and `PUT` makes sense but I was just thinking about this more and am wondering if it does.. because if the user makes a GET call then it's obvious that the returned error is regarding `GET` and same for `PUT`.

I don't have a very strong opinion about this so I will leave this up to you.

Reply:
I believe it is better to differentiate "GET" errors from "PUT" errors in our logs.